[inductor][triton] Add experimental use_tensor_descriptor config option

kundaMwiza opened the pull request:
Refactor to allow TMA descriptors to be used in general codegen. TMA descriptors can only be generated if the conditions listed in the triton documentation for [make_tensor_descriptor](https://triton-lang.org/main/python-api/generated/triton.language.make_tensor_descriptor.html) are met.

Some implementation details:
- The `TMACompatibilityChecker` class holds and checks the conditions required for a load / store operation to be represented by a tma descriptor load / store
- The current TMA API requires that the innermost block size loads atleast 16 bytes of data. e.g. if the block shape is [YBLOCK, XBLOCK] and the tensor dtype is float32, this requires that XBLOCK >= 4. It is therefore required that the triton heuristics are aware of the minimum block sizes for the IO operations in the kernel. The minimum block sizes are determined in the `TMACompatibilityChecker` class and are passed to the triton heuristics when the block sizes are not static. The heuristic config options are then filtered to ensure that the minimum block size restriction is met.

Testing:
- Refactored test_torchinductor_strided_blocks.py to also test the `use_tensor_descriptor` option.

This requires an upgrade to Triton version 3.4.0: https://github.com/pytorch/pytorch/issues/154206


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

kundaMwiza committed:
Add tests for use_tma_api and disable use of tma descriptors if load is not possible

kundaMwiza committed:
Improve comment on TMA restriction

kundaMwiza committed:
Add use tma option

kundaMwiza committed:
Fix bugs

kundaMwiza committed:
Only generate TMA loads / stores if cc >= 9

kundaMwiza committed:
Add comment

kundaMwiza committed:
Use experimental API

kundaMwiza committed:
Remove triton cpu TMA test

kundaMwiza committed:
Fix TMA load / store

kundaMwiza committed:
Refactor

kundaMwiza committed:
Keep all tma checking details inside UseTMAChecker

kundaMwiza committed:
Revert changes to copy_tests

kundaMwiza committed:
Add logging

kundaMwiza committed:
Add xfails

kundaMwiza committed:
Remove unused function

kundaMwiza committed:
Lint

kundaMwiza committed:
Update comment

kundaMwiza commented:
Needed to use `lintrunner -a -m origin/main` rather than just `lintrunner -a`

kundaMwiza committed:
Lint, improve comments and simplify

jansel commented on test/inductor/test_torchinductor_strided_blocks.py:
```diff
@@ -51,56 +53,41 @@
+# These xfails are due to the current restrictions with the TMA descriptor API.
+# see Note: TMA API Restrictions. In some cases TMA descriptors cannot be generated, and so tests
+# that assert on the expected number of descriptors (= equivalent block ptrs) will fail
+def xfail_if_use_tma_api(fn):
```
I want to confirm that these tests still run correctly (they just fall back to not using tensor descriptors).  Is that right?

kundaMwiza replied:
Yes that is correct. The tests fail due to assertion failures on the number of expected block descriptors. Currently the set values are based on block ptrs, but of course in some cases its not possible to generate a tensor descriptor. In these cases the codegen just produces the default tl.load(ptr + index)

kundaMwiza replied:
Excerpt of the failure reasons:
```
Expected 2 but got 1.
Absolute difference: 1
Relative difference: 0.5

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_multiple_max_block_non_power_of_2_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
____________________________________ TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size1_view_size1_num_block_pointers_3_num_tiles_2_cuda _____________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 694, in test_nd_tiling_odd_shapes_pointwise
    result, code = self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 3 but got 0.
Absolute difference: 3
Relative difference: 1.0

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size1_view_size1_num_block_pointers_3_num_tiles_2_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
____________________________________ TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size2_view_size2_num_block_pointers_3_num_tiles_2_cuda _____________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 694, in test_nd_tiling_odd_shapes_pointwise
    result, code = self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 3 but got 0.
Absolute difference: 3
Relative difference: 1.0

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size2_view_size2_num_block_pointers_3_num_tiles_2_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
____________________________________ TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size3_view_size3_num_block_pointers_3_num_tiles_2_cuda _____________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 694, in test_nd_tiling_odd_shapes_pointwise
    result, code = self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 3 but got 0.
Absolute difference: 3
Relative difference: 1.0

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size3_view_size3_num_block_pointers_3_num_tiles_2_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
____________________________________ TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size4_view_size4_num_block_pointers_3_num_tiles_2_cuda _____________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 694, in test_nd_tiling_odd_shapes_pointwise
    result, code = self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 3 but got 0.
Absolute difference: 3
Relative difference: 1.0

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size4_view_size4_num_block_pointers_3_num_tiles_2_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
____________________________________ TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size5_view_size5_num_block_pointers_1_num_tiles_2_cuda _____________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 694, in test_nd_tiling_odd_shapes_pointwise
    result, code = self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 1 but got 0.
Absolute difference: 1
Relative difference: 1.0

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_nd_tiling_odd_shapes_pointwise_full_size5_view_size5_num_block_pointers_1_num_tiles_2_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
_______________________________________________________________________ TritonTensorDescriptorTestCUDA.test_negative_strides_cuda _______________________________________________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor.py", line 13648, in new_test
    return value(self)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 1246, in test_negative_strides
    self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 3 but got 2.
Absolute difference: 1
Relative difference: 0.3333333333333333

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_negative_strides_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
__________________________________________________ TritonTensorDescriptorTestCUDA.test_pointwise_broadcast_nonzero_strides_prefer_nd_tiling_False_cuda __________________________________________________
Traceback (most recent call last):
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 389, in test_pointwise_broadcast_nonzero_strides
    result, (triton_code,) = self._run_and_compare(
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 169, in _run_and_compare
    count_code(self.block_descriptor_constructor_str, expected_num_block_pointers)
  File "/home/ubuntu/pytorch/test/inductor/test_torchinductor_strided_blocks.py", line 164, in count_code
    self.assertEqual(count, expected)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 4168, in assertEqual
    raise error_metas.pop()[0].to_error(  # type: ignore[index]
AssertionError: Scalars are not equal!

Expected 3 but got 2.
Absolute difference: 1
Relative difference: 0.3333333333333333

To execute this test, run the following from the base repo dir:
    python test/inductor/test_torchinductor_strided_blocks.py TritonTensorDescriptorTestCUDA.test_pointwise_broadcast_nonzero_strides_prefer_nd_tiling_False_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
```

jansel commented on torch/_inductor/config.py:
```diff
@@ -1280,6 +1280,19 @@ class triton:
+    #   assume_aligned_inputs to also be enabled
+    # TMA descriptors are only going to be generated if the above conditions
+    # can be satisfied, along with any existing requirements for index expressions
+    use_tma_api = False
```
```suggestion
    use_tensor_descriptor = False
```

jansel reviewed (changes requested):
lint errors

kundaMwiza committed:
- Lint
- rename use_tma_api -> use_tensor_descriptor
- check blocks for persistent_reduction when block params are available

kundaMwiza committed:
Rename TMADescriptorOptions->TensorDescriptorOptions

kundaMwiza committed:
Rename UseTMAChecker -> TMACompatibilityChecker

kundaMwiza committed:
TritonTMADescriptorTestCUDA->TritonTensorDescriptorTestCUDA

kundaMwiza committed:
Fix use str instead of sympy variable

kundaMwiza committed:
Remove tensor descriptor lines as they are not generated

kundaMwiza committed:
Log persistent reduction tma compat failure

jansel commented:
@pytorchbot merge
